### PR TITLE
Browser Recorder Blog

### PR DESCRIPTION
**Pull request by aryangupta701**

### CODE REVIEW

File: site/content/blog/2023-09-09-browser-extension/index.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+---
+title: "Parsing .DS_Store files with ZAP"
+summary: "ZAP Spider can now probe and parse macOS' .DS_Store files."
```
Comment (thc202) [inline site/content/blog/2023-09-09-browser-extension/index.md:3]: Needs update.

File: site/content/blog/2023-09-09-browser-extension/index.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+images:
+- https://www.zaproxy.org/blog/2023-09-09-browser-extension/images/zapgsoc.png
+type: post
+tags:
+- blog
+- browser extension
+- gsoc 2023
+date: "2023-09-09"
```
Comment (psiinon) [inline site/content/blog/2023-09-09-browser-extension/index.md:11]: OK to publish on Monday? Will give us more time to review and publish the client add-on.
I would expect it to get more visibility during a weekday as well 😁

**Comment by psiinon** [comment]
Definitely need to explain to people how they can try this out for themselves.
This means installing the client add-on from the Marketplace (we can provide the link), noting it only works with Firefox (for now) and that they have to approve for each site.
I think it will fail on Chrome as the latest version has not been approved yet? If so should we disable Chrome support in Zest until we can update it?

File: site/content/blog/2023-09-09-browser-extension/index.md (hunk 1 above)
Comment (psiinon) [reply site/content/blog/2023-09-09-browser-extension/index.md:3]: "GSoC 2023 Browser Recorder" (or similar) might be a better blog title than "Browser Recorder - Exploring My Google Summer of Code 2023 Journey" - its better to keep them shorter (IMO). You can still use that title lower down..

File: site/content/blog/2023-09-09-browser-extension/images/zapgsoc.png
Comment (psiinon) [inline site/content/blog/2023-09-09-browser-extension/images/zapgsoc.png:1]: Trim some of the whitespace around the images?

File: site/content/blog/2023-09-09-browser-extension/index.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+2. Typing
+3. Frame Switches 
+
+Here is how the browser extension looks like:
+
+![browser-recoder](images/browser-recorder.png)
+
+## Client Add On
+
+The client add belongs to the ZAP extensions which helps to integrate with the browser extension. The client add on is responsible for the following tasks:
```
Comment (psiinon) [inline site/content/blog/2023-09-09-browser-extension/index.md:59]: "Client add-on" and link to the relevant web page (we'll need to provide that in a bit)

File: site/content/blog/2023-09-09-browser-extension/index.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+* Responsible for the communication between the browser extension and the ZAP. 
+* Used by the Zest Addon to record the user interactions and generate the Zest script.
+
+## Integration with Zest AddOn 
+
+The Zest Addon is extended to provide a new record type called `Client (browser) side script`. This record type is used to record the user interactions and generate the Zest script. The Zest Addon leverages the power of the client add on and the browser extension to record the user interactions and generate the Zest script.
```
Comment (psiinon) [inline site/content/blog/2023-09-09-browser-extension/index.md:66]: "Zest add-on" and link to https://www.zaproxy.org/docs/desktop/addons/zest/
I think it should always be "add-on" - lower case with a '-' might need to update a few other places..

**Review by psiinon — COMMENTED** [review]
Some of the titles are very "wordy" - I'd be tempted to make them shorter, but in the end thats up to you.

File: site/content/blog/2023-09-09-browser-extension/index.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+---
+# Browser Recorder - Exploring My Google Summer of Code 2023 Journey
+
+![ZAP GSoC Image](images/zapgsoc.png)
+
+## My Introduction
+
+I'm Aryan Gupta, a passionate software developer with a knack for problem-solving. I've worked with various programming languages, including Java, JavaScript, TypeScript, C++, and Python. As a seasoned full-stack developer well-versed in the MERN stack, I've always been eager to learn new things. My curiosity also extends to cybersecurity, where I'm fascinated by the challenge of uncovering online vulnerabilities.
+
+Contact Me Here: [GitHub](https://www,github.com/aryangupta701) | [LinkedIn](https://www.linkedin.com/in/aryan-gupta-78273a1b6/) | [Twitter](https://twitter.com/aryangupta_701)
```
Comment (psiinon) [inline site/content/blog/2023-09-09-browser-extension/index.md:23]: This info would be better on a Team page.
You can add a page for yourself here: https://www.zaproxy.org/docs/team/ (after Core Team, before Extended). You can do that in another PR which would get merged first if you like 😁

File: site/content/blog/2023-09-09-browser-extension/index.md (hunk 2 above)
Comment (aryangupta701) [reply site/content/blog/2023-09-09-browser-extension/index.md:11]: Yes sure

File: site/content/blog/2023-09-09-browser-extension/index.md (hunk 5 above)
Comment (aryangupta701) [reply site/content/blog/2023-09-09-browser-extension/index.md:23]: okay

**Comment by aryangupta701** [comment]
ready for review.I think images are not looking good. Is there any way to modify these form md file directly? I wanted to align it in center and create a border too.

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
+---
+title: "GSoC 2023 Browser Recorder"
+summary: "ZAP has introduced a new feature to record pre-task activities such as logging in etc. using Browser Recorder."
+images:
+- https://www.zaproxy.org/blog/2023-09-11-browser-extension/images/zapgsoc.png
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:5]: The link needs to be updated, `extension` → `recorder`.

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+The proposed solution is simple yet powerful: a browser extension that records user interactions while navigating a website. These interactions are then intelligently transformed into a Zest script, a human-readable format used in ZAP. This script is generated in JSON format, ensuring compatibility and ease of use with ZAP's existing infrastructure. This innovative approach eliminates the need for manual scripting, allowing users to focus solely on testing for vulnerabilities. 
+
+## The ZAP Browser Recorder Project: A Closer Look
+
+The ZAP Browser Recorder project is comprised of two main components: the browser extension and the client add on. The browser extension is responsible for recording user interactions, while the client add on facilitates communication between the browser extension and ZAP. Together, these components form a cohesive solution that streamlines the website testing process.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:40]: For the first mention `[Client Side Integration add-on](/docs/desktop/addons/client-side-integration/)` and for the following ones `Client add-on` (capitalize client).

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+2. Typing
+3. Frame Switches 
+
+Here is how the browser extension looks like:
+
+![browser-recoder](images/browser-recorder.png)
+
+## Client add-on
+
+The [client add-on](https://www.zaproxy.org/docs/desktop/addons/client/) belongs to the ZAP extensions which helps to integrate with the browser extension. The client add-on is responsible for the following tasks:
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:54]: - See previous comment.
 - What does it mean "belongs to the ZAP extensions", the repository?

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+* Responsible for the communication between the browser extension and the ZAP.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:56]: `and ZAP`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+Here is how the browser extension looks like:
+
+![browser-recoder](images/browser-recorder.png)
+
+## Client add-on
+
+The [client add-on](https://www.zaproxy.org/docs/desktop/addons/client/) belongs to the ZAP extensions which helps to integrate with the browser extension. The client add-on is responsible for the following tasks:
+
+* Responsible for the communication between the browser extension and the ZAP. 
+* Used by the Zest add-on to record the user interactions and generate the Zest script.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:57]: I think they should be linked when first mentioned (though does not hurt linking in several places).

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+## Integration with Zest add-on 
+
+The [Zest add-on](https://www.zaproxy.org/docs/desktop/addons/zest/) is extended to provide a new record type called `Client (browser) side script`. This record type is used to record the user interactions and generate the Zest script. The Zest add-on leverages the power of the client add-on and the browser extension to record the user interactions and generate the Zest script.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:61]: `(/docs/desktop/addons/zest/)`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+The ZAP Browser Recorder's first verison has been released and can be used by the users. We have released the browser extension for the following browsers:
+
+* Chrome (Under Review by the Chrome Web Store)
+* Firefox (Published)
+
+The browser recorder feature can be used in two ways: 
+
+### 1. Using the Zest add-on
+
+- First of all you need to install the client and zest add-on from the ZAP marketplace.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:83]: `Client`, `Zest`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+* Chrome (Under Review by the Chrome Web Store)
+* Firefox (Published)
+
+The browser recorder feature can be used in two ways: 
+
+### 1. Using the Zest add-on
+
+- First of all you need to install the client and zest add-on from the ZAP marketplace. 
+- Then you need to make sure that the browser extension has been enabled for chrome as we currently only support integration with chrome browser.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:84]: `Chrome`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+The browser recorder feature can be used in two ways: 
+
+### 1. Using the Zest add-on
+
+- First of all you need to install the client and zest add-on from the ZAP marketplace. 
+- Then you need to make sure that the browser extension has been enabled for chrome as we currently only support integration with chrome browser. 
+- Then you will have to go to the `Record a Zest Script` option as shown in the image below:
+
+![Zest Record Dialoge](images/zest-record.png)
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:87]: `Zest Record Button`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+- Then you will have to select the `Client (browser) side script` option and fill the initial URL which you wish to record as shown in the image below:
+
+![Zest Record Dialoge](images/zest-record-dialog.png)
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:91]: `Record Zest Script Dialog`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+- Now, click on start start recording. The browser will open and you will be redirected to the URL to have entered. Now click on the extensions and you will see the ZAP browser extension.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:93]: - `on Start Recording`.
- `you have entered`?

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+- Then you will have to go to the `Record a Zest Script` option as shown in the image below:
+
+![Zest Record Dialoge](images/zest-record.png)
+
+- Then you will have to select the `Client (browser) side script` option and fill the initial URL which you wish to record as shown in the image below:
+
+![Zest Record Dialoge](images/zest-record-dialog.png)
+
+- Now, click on start start recording. The browser will open and you will be redirected to the URL to have entered. Now click on the extensions and you will see the ZAP browser extension. 
+- Click on the extension and you will se a popup window. Now click on the `Record` button to start the recording.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:94]: `see a`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+- Now you can do whatever you want to record and when you are done with the recording you can click on the `Stop` button.
+- The script will be automatically generated by the Zest add-on and you can see the script in the Scripts -> Standalone Scripts section.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:96]: - `Stand Alone`
- Worth noting that the default type selected is Session Management not Stand Alone, the screenshot shown in the blog post is also showing the Session Management type selected.

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+![Zest Record Dialoge](images/zest-record-dialog.png)
+
+- Now, click on start start recording. The browser will open and you will be redirected to the URL to have entered. Now click on the extensions and you will see the ZAP browser extension. 
+- Click on the extension and you will se a popup window. Now click on the `Record` button to start the recording.
+- Now you can do whatever you want to record and when you are done with the recording you can click on the `Stop` button.
+- The script will be automatically generated by the Zest add-on and you can see the script in the Scripts -> Standalone Scripts section. 
+
+### 2. Using the Browser Extension
+
+- First of all you need to install the browser extension from the chrome/firefox web store.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:100]: - Chrome/Firefox
- Might be a good idea to suggest the use of launched browsers, which will already have the extension.

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+- You will see the browser extension icon in the top right corner of the browser as shown in the image below:
+
+![Browser Extension Icon](images/zap-browser-extension-icon.png)
+
+- When you click on the browser extension icon you will see a popup window as shown in the image below: 
+
+![Browser Extension Popup](images/zap-browser-extension-popup.png)
+
+- Before starting the recording go to settings icon and make sure that the `enable` option is turned off (disabling this will prevent from unnecessarily recording the server side recording). You donot need to do this step if you record using the Zest add-on.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:109]: `You do not`?

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+
+- When you click on the browser extension icon you will see a popup window as shown in the image below: 
+
+![Browser Extension Popup](images/zap-browser-extension-popup.png)
+
+- Before starting the recording go to settings icon and make sure that the `enable` option is turned off (disabling this will prevent from unnecessarily recording the server side recording). You donot need to do this step if you record using the Zest add-on. 
+
+- Then you can start the recording by clicking on the `Record` button.
+- Now you can do whatever you want to record and when you are done with the recording you can click on the `Stop` button.
+- Then you can click on the `Download` button to download the zest script.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:113]: `Zest`

File: site/content/blog/2023-09-11-browser-recorder/index.md (hunk 18 above)
Comment (aryangupta701) [reply site/content/blog/2023-09-11-browser-recorder/index.md:96]: Oh yes thank you for pointing that out.

File: site/content/blog/2023-09-11-browser-recorder/index.md (hunk 8 above)
Comment (aryangupta701) [reply site/content/blog/2023-09-11-browser-recorder/index.md:54]: yes the ZAP extensions but i think add-on already conveys that message

File: site/content/blog/2023-09-11-browser-recorder/index.md (hunk 19 above)
Comment (aryangupta701) [reply site/content/blog/2023-09-11-browser-recorder/index.md:100]: didn't understand the second point can you please explain

Comment (thc202) [reply site/content/blog/2023-09-11-browser-recorder/index.md]: Just mention that the user can launch the browsers from ZAP (and link to quick start docs).

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+
+The proposed solution is simple yet powerful: a browser extension that records user interactions while navigating a website. These interactions are then intelligently transformed into a Zest script, a human-readable format used in ZAP. This script is generated in JSON format, ensuring compatibility and ease of use with ZAP's existing infrastructure. This innovative approach eliminates the need for manual scripting, allowing users to focus solely on testing for vulnerabilities. 
+
+## The ZAP Browser Recorder Project: A Closer Look
+
+The ZAP Browser Recorder project is comprised of two main components: the browser extension and the [Client Side Integration add-on](/docs/desktop/addons/client-side-integration/). The browser extension is responsible for recording user interactions, while the Client add on facilitates communication between the browser extension and ZAP. Together, these components form a cohesive solution that streamlines the website testing process.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:40]: `add-on`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+Here is how the browser extension looks like:
+
+![browser-recoder](images/browser-recorder.png)
+
+## Client Side Integration add-on
+
+The [Client Side Integration add-on](/docs/desktop/addons/client-side-integration/) helps to integrate with the browser extension. The Client add-on is responsible for the following tasks:
+
+* Responsible for the communication between the browser extension and ZAP. 
+* Used by the [Zest add-on](https://www.zaproxy.org/docs/desktop/addons/zest/) to record the user interactions and generate the Zest script.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:57]: Remove the domain, for own links it's better to not include the domain so they point to the running server (e.g. testing locally).

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+
+## Integration with Zest add-on 
+
+The [Zest add-on](/docs/desktop/addons/zest/) is extended to provide a new record type called `Client (browser) side script`. This record type is used to record the user interactions and generate the Zest script. The Zest add-on leverages the power of the client add-on and the browser extension to record the user interactions and generate the Zest script.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:61]: `Client`

File: site/content/blog/2023-09-11-browser-recorder/index.md (hunk 19 above)
Comment (aryangupta701) [reply site/content/blog/2023-09-11-browser-recorder/index.md]: isn't this resolved yet?

Comment (thc202) [reply site/content/blog/2023-09-11-browser-recorder/index.md]: Probably, I didn't check the latest change.

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+![Record Zest Script Dialog](images/zest-record-dialog.png)
+
+- Now, click on Start Recording. The browser will open and you will be redirected to the URL you have entered. Now click on the extensions and you will see the ZAP browser extension. 
+- Click on the extension and you will see a popup window. Now click on the `Record` button to start the recording.
+- Now you can do whatever you want to record and when you are done with the recording you can click on the `Stop` button.
+- The script will be automatically generated by the Zest add-on and you can see the script in the Scripts -> Session Management Scripts section. 
+
+### 2. Using the Browser Extension
+
+- For ease, users can use the existing ZAP browsers from the [Quick Start](/docs/desktop/addons/quick-start/) section which will already have extension install.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:100]: `… can launch browsers from ZAP using the [Quick Start add-on](…/#manual-explore), which will already have the extension installed.`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+
+The proposed solution is simple yet powerful: a browser extension that records user interactions while navigating a website. These interactions are then intelligently transformed into a Zest script, a human-readable format used in ZAP. This script is generated in JSON format, ensuring compatibility and ease of use with ZAP's existing infrastructure. This innovative approach eliminates the need for manual scripting, allowing users to focus solely on testing for vulnerabilities. 
+
+## The ZAP Browser Recorder Project: A Closer Look
+
+The ZAP Browser Recorder project is comprised of two main components: the browser extension and the [Client Side Integration add-on](/docs/desktop/addons/client-side-integration/). The browser extension is responsible for recording user interactions, while the Client add-on facilitates communication between the browser extension and ZAP. Together, these components form a cohesive solution that streamlines the website testing process. 
+
+We have submitted a newer version of the extension with Chrome Web Store and it is currently under review.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:42]: `extension to the Chrome Web Store`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+
+* Chrome (Latest version under review by the Chrome Web Store)
+* Firefox (Published)
+
+The browser recorder feature can be used in two ways: 
+
+### 1. Using the Zest add-on
+
+- First of all you need to install the Client and Zest add-on from the ZAP marketplace. 
+- Then you need to make sure that the browser extension has been enabled for Chrome as we currently only support integration with Chrome browser because of the permission issues in Firefox. We are working on to provide support for firefox as well.
```
Comment (thc202) [inline site/content/blog/2023-09-11-browser-recorder/index.md:86]: `on providing full support for Firefox`

File: site/content/blog/2023-09-11-browser-recorder/index.md [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+
+## Introducing Browser Recorder Project
+
+In the ever-evolving landscape of web development and cybersecurity, the importance of comprehensive testing cannot be overstated. However, testing websites often involves repetitive pre-task activities like logging in, which can be time-consuming and cumbersome. Recognizing this challenge, the ZAP Browser Recorder project sets out to revolutionize the way websites are tested by addressing the inconvenience users face during these initial steps.
+
+## The Solution
+
+Enter the ZAP Browser Recorder project - a visionary solution to this often-overlooked challenge. The project's primary goal is to streamline the website testing process, making it more efficient and user-friendly. At its core, the ZAP Browser Recorder aims to provide a seamless bridge between user interactions and ZAP's capabilities.
```
Comment (psiinon) [inline site/content/blog/2023-09-11-browser-recorder/index.md:34]: "visionary" - thats probably stretching it a bit 😉 There are other browser recorders out there...

**Review by psiinon — APPROVED** [review]

**Review by thc202 — APPROVED** [review]

**Pull request merged**